src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs

```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
             });
         }
 
-        private void PartitionsAssigned(IEnumerable<TopicPartition> partitions)
+        private class FlushMessagesOfRevokedPartitionsHandler : IPartitionEventHandler
         {
-            TopicPartitions = TopicPartitions.Union(partitions);
-            Log.Debug("[{0}] Partitions were assigned: {1}", ConsumerActor.Path.Name, string.Join(", ", partitions));
-            RequestMessages();
+            private IImmutableSet<TopicPartitionOffset> _lastRevoked = ImmutableHashSet<TopicPartitionOffset>.Empty;
+            private readonly SingleSourceStageLogic<K, V, TMessage> _stageLogic;
+
+            public FlushMessagesOfRevokedPartitionsHandler(SingleSourceStageLogic<K, V, TMessage> stageLogic)
+            {
+                _stageLogic = stageLogic;
+            }
+
+            public void OnRevoke(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions,
+                IRestrictedConsumer consumer)
+            {
+                _lastRevoked = revokedTopicPartitions;
+            }
+
+            public void OnLost(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                _stageLogic.FilterRevokedPartitionAsyncCallback(revokedTopicPartitions);
+            }
+
+            public void OnAssign(IImmutableSet<TopicPartition> assignedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                // remove all of our previous revoked partitions that are not in the new assignment
+                _stageLogic.FilterRevokedPartitionAsyncCallback(_lastRevoked
```

Comment:
Revoke messages from all partitions that were not assigned to us - when the Kafka broker rebalances partitions all partitions get revoked from every body, but one of the original partitions we owned might get assigned back to us again, so we don't want to discard messages until we're sure about which partitions we're keeping.